subtests/docker_cli/login/login.py

```diff
@@ -0,0 +1,333 @@
+        self.sub_stuff['port'] = port
+        self.sub_stuff['server'] = localhost + ':' + port
+
+    def _create_htpasswd(self):
+        """
+        Write an htpasswd file in our current directory, for use by
+        the registry container (via bind mount). Use the registry image
+        itself to run htpasswd, with -n so output goes to stdout, then
+        capture that stdout and write it to a file in our current directory.
+        """
+        self.sub_stuff['user'] = 'testuser'
+        self.sub_stuff['passwd'] = utils.generate_random_string(12)
+
+        registry_fqin = self.config['registry_fqin']
+        subargs = ['--rm', '--entrypoint', 'htpasswd', registry_fqin,
+                   '-Bbn', self.sub_stuff['user'], self.sub_stuff['passwd']]
+        htpasswd_cmdresult = DockerCmd(self, 'run', subargs).execute()
+        with open('htpasswd', 'wb') as htpasswd_fh:
+            htpasswd_fh.write(htpasswd_cmdresult.stdout)
+
+    def _preserve_docker_auth_file(self):
+        """
+        docker login creates/overwrites a standard file; preserve it
+        if it already exists.
+        """
+        auth_file = self.config['docker_auth_file']
+        if not os.path.exists(auth_file):
+            return
+        saved_path = auth_file + '.PRESERVED'
+        os.rename(auth_file, saved_path)
+        self.sub_stuff['docker_auth_file_preserved'] = saved_path
+
+    def _restore_docker_auth_file(self):
+        """
+        Restore previous docker auth file, if any. Otherwise, remove.
+        """
+        auth_file = self.config['docker_auth_file']
+        key = 'docker_auth_file_preserved'
+        if key in self.sub_stuff:
+            os.rename(self.sub_stuff[key], auth_file)
+        else:
+            os.unlink(auth_file)
+
+    def _wait_for_registry_ready(self, timeout=None):
+        """
+        Registry container takes some time to spin up and be ready. This
+        can take a few seconds, and we don't want to run login commands
+        before it's ready. We can't just connect to the port, because
+        docker opens it immediately. Best solution seems to be to GET
+        a well-defined v2 URL in the registry. It will fail with SSLEOF
+        at the beginning, then fail with HTTP 401 (Unauthorized). Once
+        we get that, trust that it's ready for login.
+        """
+        if timeout is None:
+            timeout = 10
+
+        url = "https://{server}/v2/".format(**self.sub_stuff)
+        end_time = time.time() + timeout
+        while time.time() <= end_time:
+            try:
+                urllib2.urlopen(url)
+            except urllib2.HTTPError as e:
+                if e.reason == 'Unauthorized':
+                    return
+            except urllib2.URLError as e:
+                # FIXME: assume that it's SSLEOFError; how can we make sure?
+                pass
+            time.sleep(0.5)
+        raise DockerTestFail("timed out waiting for registry")
```

Comment:
Totally, just kill the fixme.  There are way too many things that can go wrong w/ most things involving networking.  Simply checking for success / !success is fine, and logging as many details as is practical.  Most of the time when it breaks, it's because someone fudged a name/ip or the server is just down.